src/machine/machine_atmega.go

```diff
@@ -155,3 +157,109 @@ func (uart UART) WriteByte(c byte) error {
 	avr.UDR0.Set(c) // send char
 	return nil
 }
+
+// SPIConfig is used to store config info for SPI.
+type SPIConfig struct {
+	Frequency uint32
+	LSBFirst  bool
+	Mode      uint8
+	CSPins    []Pin
+}
+
+// SPI is for the Serial Peripheral Interface
+// Data is taken from http://ww1.microchip.com/downloads/en/DeviceDoc/ATmega48A-PA-88A-PA-168A-PA-328-P-DS-DS40002061A.pdf page 169 and following
+type SPI struct {
+	// The registers for the SPIx port set by the chip
+	spcr *volatile.Register8
+	spdr *volatile.Register8
+	spsr *volatile.Register8
+
+	// The io pins for the SPIx port set by the chip
+	sck Pin
+	sdi Pin
+	sdo Pin
+}
+
+// Configure is intended to setup the SPI interface.
+func (s SPI) Configure(config SPIConfig) error {
+
+	// This is only here to help catch a bug with the configuration
+	// where a machine missed a value.
+	if s.spcr == (*volatile.Register8)(unsafe.Pointer(uintptr(0))) ||
+		s.spsr == (*volatile.Register8)(unsafe.Pointer(uintptr(0))) ||
+		s.spdr == (*volatile.Register8)(unsafe.Pointer(uintptr(0))) ||
+		s.sck == 0 || s.sdi == 0 || s.sdo == 0 {
+		return errSPIInvalidMachineConfig
+	}
+
+	// Make the defaults meaningful
+	if config.Frequency == 0 {
+		config.Frequency = 4000000
+	}
+
+	// Default all port configuration bits to 0 for simplicity
+	s.spcr.Set(0)
+	s.spsr.Set(0)
+
+	// Setup pins output configuration
+	s.sck.Configure(PinConfig{Mode: PinOutput})
+	s.sdi.Configure(PinConfig{Mode: PinInput})
+	s.sdo.Configure(PinConfig{Mode: PinOutput})
+
+	// Ensure the CS pins are all high before enabling them to prevent any
+	// momentary low glitches that can confuse the other chips
+	for _, cs := range config.CSPins {
```

Comment:
I do not really see what benefit having the `CSPins` slice provides here. Especially since it is not used anywhere else.

Reply:
It's not included elsewhere but some place needs to be the first :).  (There isn't a governing interface for SPI that I could see).

I'll argue that the CS pins are missing from the other impls and not the other way around.  Here are a few examples of code that all have a bug on Atmega chips that may or may not have issues on different chips:

- https://github.com/tinygo-org/drivers/blob/e376785596dc8269f3e8aa42a9bf75fb1457febc/wifinina/wifinina.go#L271
- https://github.com/tinygo-org/drivers/blob/e376785596dc8269f3e8aa42a9bf75fb1457febc/flash/transport_spi.go#L44
- https://github.com/tinygo-org/drivers/blob/e376785596dc8269f3e8aa42a9bf75fb1457febc/bmi160/bmi160.go#L31

The output pin goes low for a few ns & then high.  For some chips this can signal the start of a transaction & inadvertently cause the first few attempts to use a chip not work.

I definitely agree with standardizing the SPI port usage.  I'm hopeful that driver or application code can delegate pin configuration to the SPI code for the chip (like it does for the other SPI pins).

But if this is the only thing keeping this code from being merged I'll remove it.

Reply:
@deadprogram If the only thing blocking this merge is the CS handling - I'll remove it to get the rest of the code in.  I'm fine dealing with it in a later PR as well.  I'm only pinging you here because there's so many comments & I don't want you to have to waste your time looking through them if this is the only blocker.

Reply:
I think we should probably not try to make changes to the interface for all SPI as part of this PR.

That would better served as part of a broader conversation of possible changes to the SPI interface.